File: wrappers/dotnet/src/BbsSignatures/Keys/BlsKeyPair.cs
using System.Collections.ObjectModel;
using System.Text;

namespace BbsSignatures
{
    /// <summary>
    /// A BLS 12-381 key pair
    /// </summary>
    public class BlsKeyPair
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="BlsKeyPair" /> class.
        /// </summary>
        /// <param name="secretKey">The secret key.</param>
        /// <param name="deterministicPublicKey">The deterministic public key.</param>
        public BlsKeyPair(byte[] publicKey, byte[]? secretKey = default)
        {
            SecretKey = secretKey;
            PublicKey = publicKey;
        }

        /// <summary>
        /// Default BLS 12-381 public key length
        /// </summary>
        public static int PublicKeyG1Size => NativeMethods.bls_public_key_g1_size();

        /// <summary>
        /// Default BLS 12-381 public key length
        /// </summary>
        public static int PublicKeyG2Size => NativeMethods.bls_public_key_g2_size();

        /// <summary>
        /// Default BLS 12-381 private key length
        /// </summary>
        public static int SecretKeySize => NativeMethods.bls_secret_key_size();

        /// <summary>
        /// Raw public key value for the key pair
        /// </summary>
        /// <returns></returns>
        public byte[] PublicKey { get; internal set; }

        /// <summary>
        /// Raw secret/private key value for the key pair
        /// </summary>
        /// <value>
        /// The key.
        /// </value>
        public byte[]? SecretKey { get; internal set; }

        /// <summary>
        /// Generates new BBS+ public key from the current BLS12-381
        /// </summary>
        /// <param name="messageCount">The message count.</param>
        /// <returns></returns>
        public BbsKey GetBbsKey(uint messageCount)
        {
            using var context = new UnmanagedMemory();

            if (SecretKey != null)
            {
                NativeMethods.bls_secret_key_to_bbs_key(context.ToBuffer(SecretKey), messageCount, out var publicKey, out var error);
                context.ThrowOnError(error);

                return new BbsKey(context.ToByteArray(publicKey), messageCount);
            }
            else if (IsG2())
            {
                NativeMethods.bls_public_key_to_bbs_key(context.ToBuffer(PublicKey), messageCount, out var publicKey, out var error);
                context.ThrowOnError(error);

                return new BbsKey(context.ToByteArray(publicKey), messageCount);
            }
            throw new BbsException("Cannot generate BbsKey from G1 public key");
        }

        /// <summary>
        /// Returns <c>true</c> if the current instance is of G2 cyclic group, otherwise <c>false</c>
        /// </summary>
        /// <returns></returns>
        public bool IsG2() => PublicKey.Length == PublicKeyG2Size;

        /// <summary>
        /// Returns <c>true</c> if the current instance is of G1 cyclic group, otherwise <c>false</c>
        /// </summary>
        /// <returns></returns>
        public bool IsG1() => PublicKey.Length == PublicKeyG1Size;

        /// <summary>
        /// Creates new <see cref="BlsKeyPair"/> using a input seed as string.
        /// </summary>
        /// <param name="seed">The seed.</param>
        /// <returns></returns>
        public static BlsKeyPair GenerateG1(string? seed = null)
        {
            using var context = new UnmanagedMemory();

            var result = NativeMethods.bls_generate_g1_key(
                seed is null ? ByteBuffer.None : context.ToBuffer(Encoding.UTF8.GetBytes(seed)),
                out var publicKey,
                out var secretKey,
                out var error);

            context.ThrowOnError(error);

            return new BlsKeyPair(context.ToByteArray(publicKey), context.ToByteArray(secretKey));
        }

        /// <summary>
        /// Creates new <see cref="BlsKeyPair"/> using a input seed as string.
        /// </summary>
        /// <param name="seed">The seed.</param>
        /// <returns></returns>
        public static BlsKeyPair GenerateG2(string? seed = null)
        {
            using var context = new UnmanagedMemory();

            var result = NativeMethods.bls_generate_g2_key(
                seed is null ? ByteBuffer.None : context.ToBuffer(Encoding.UTF8.GetBytes(seed)),
                out var publicKey,
                out var secretKey,
                out var error);

            context.ThrowOnError(error);

            return new BlsKeyPair(context.ToByteArray(publicKey), context.ToByteArray(secretKey));
        }
    }
}

File: wrappers/python/ursa_bbs_signatures/models/keys/BlsKeyPair.py
from typing import Optional, Union

from .BbsKey import BbsKey
from ..._ffi.bindings.bls import (
    bls_public_key_g1_size,
    bls_public_key_g2_size,
    bls_secret_key_to_bbs_key,
    bls_public_key_to_bbs_key,
    bls_generate_g1_key,
    bls_generate_g2_key,
    bls_get_public_key, bls_secret_key_size
)
from ...models.BbsException import BbsException


class BlsKeyPair:
    def __init__(self, public_key: bytes, secret_key: Optional[bytes] = None) -> None:
        """ A BLS 12-381 key pair

        Args:
            public_key: The deterministic public key
            secret_key: The secret key
        """
        self.public_key = public_key
        self.secret_key = secret_key

    @staticmethod
    def public_g1_key_size() -> int:
        return bls_public_key_g1_size()

    @staticmethod
    def public_g2_key_size() -> int:
        return bls_public_key_g2_size()

    @staticmethod
    def secret_key_size() -> int:
        return bls_secret_key_size()

    @property
    def is_g1(self) -> bool:
        return len(self.public_key) == bls_public_key_g1_size()

    @property
    def is_g2(self) -> bool:
        return len(self.public_key) == bls_public_key_g2_size()

    def get_bbs_key(self, message_count: int) -> BbsKey:
        if self.secret_key:
            pub_key = bls_secret_key_to_bbs_key(self.secret_key, message_count)
            return BbsKey(bytes(pub_key), message_count)
        elif self.is_g2:
            pub_key = bls_public_key_to_bbs_key(self.public_key, message_count)
            return BbsKey(bytes(pub_key), message_count)

        raise BbsException("Cannot generate BbsKey from G1 public key")

    @classmethod
    def generate_g1(cls, seed: Optional[Union[str, bytes]] = None) -> "BlsKeyPair":
        res = bls_generate_g1_key(seed)

        return cls(res["public_key"], res["secret_key"])

    @classmethod
    def generate_g2(cls, seed: Optional[Union[str, bytes]] = None) -> "BlsKeyPair":
        res = bls_generate_g2_key(seed)

        return cls(res["public_key"], res["secret_key"])

    @classmethod
    def from_secret_key(cls, secret_key: bytes) -> "BlsKeyPair":
        pub_key = bls_get_public_key(secret_key)
        return cls(pub_key, secret_key)
